### PR TITLE
Multi-build direct dependency support + build all ordering

### PR DESCRIPTION
**Pull request by samuel-williams-da**
Fixes issues listed [here](https://discuss.daml.com/t/how-is-the-order-computed-for-building-models-in-daml-build-all/7195)
Please also approve [the backport](https://github.com/digital-asset/daml/pull/18957)

### CODE REVIEW

**Review by remyhaemmerle-da — APPROVED** [review]
LGTM

File: sdk/compiler/damlc/lib/DA/Cli/Damlc.hs [hunk 1]
```diff
@@ -1045,7 +1045,9 @@ multiPackageBuildEffect relativize mPkgConfig multiPackageConfig projectOpts opt
     IDE.shutdown
     $ \ideState -> runActionSync ideState
       $ case mRootPkgData of
-          Nothing -> void $ uses_ BuildMulti $ toNormalizedFilePath' <$> mpPackagePaths multiPackageConfig
+          -- `uses_` tries to run in "parallel", but ends up just running backwards for multi-build.
+          -- We use `reverse` here to keep the build order consistent with the multi-package.yaml
+          Nothing -> void $ uses_ BuildMulti $ reverse $ toNormalizedFilePath' <$> mpPackagePaths multiPackageConfig
```
Comment (remyhaemmerle-da) [inline sdk/compiler/damlc/lib/DA/Cli/Damlc.hs:1050]: Naive question.
Why do we need the `reverse` here ?

Comment (samuel-williams-da) [reply sdk/compiler/damlc/lib/DA/Cli/Damlc.hs:1050]: Its explained in the comment above :)

Comment (remyhaemmerle-da) [reply sdk/compiler/damlc/lib/DA/Cli/Damlc.hs:1050]: Thanks

**Pull request merged**